src/passes/function_passes/CheckPointPass.cpp

```diff
@@ -0,0 +1,200 @@
+                mod, Intrinsic::experimental_patchpoint_void);
+          } else {
+            // The function is __unopt_trace -> insert a stackmap call instead
+            // of a patchpoint call
+            intrinsic = Intrinsic::getDeclaration(
+                mod, Intrinsic::experimental_stackmap);
+          }
+          // Pass the live locations to the stackmap/patchpoint call. This also
+          // inserts the size of each 'live' location into the stackmap.
+          for (BasicBlock::iterator prev_inst = bb.begin(); prev_inst != it;
+               ++prev_inst) {
+            // XXX need to accurately identify the live registers - this
+            // currently assumes each instruction in the current basic block to
+            // represent a live location.
+            if (prev_inst != it &&
+                  prev_inst->use_begin() != prev_inst->use_end()) {
+              // Only look at instructions which produce values that have at
+              // least one use. If a result is not used anywhere, it is going
+              // to be removed by future optimization passes, so it should not
+              // be recorded.
+              args.push_back(&*prev_inst);
+              auto size_of_inst = builder.getInt64(8); // XXX default size
+              if (isa<AllocaInst>(prev_inst)) {
+                Type *t = cast<AllocaInst>(*prev_inst).getAllocatedType();
+                size_of_inst = builder.getInt64(
+                    data_layout.getTypeAllocSize(t));
+              }
+              // also insert the size of the recorded location to know how
+              // many bytes to copy at runtime
+              args.push_back(size_of_inst);
+            }
+          }
+          auto call_inst = builder.CreateCall(intrinsic, args);
+          ++call_inst_count;
+        }
+      }
+    }
+    if (start_sm_id.find(twin_fun) == start_sm_id.end()) {
+      // The 'twin' of the function has not yet been processed, which means
+      // sm_id was used as a 'base' ID, so it needs to be updated, since
+      // future functions can no longer use it as a 'base'.
+      sm_id = start_sm_id[fun_name] + call_inst_count;
+    }
+    return true;
+  }
+
+  /*
+   * Return the name of 'twin' function of function with the specified name.
+   *
+   * If name is the name of an optimized function, then the name of the
+   * function's 'twin' is the name prefixed with '__unopt_'. If the specified
+   * name is the name of an unoptimized function, then this returns the name
+   * of the optimized function (it strips off the '__unopt_' prefix).
+   *
+   * "fun" -> "__unopt_fun" or "__unopt_fun" -> "fun"
+   */
+  static StringRef getTwinName(StringRef name) {
+    if (name.startswith(UNOPT_PREFIX)) {
+      return name.substr(StringRef(UNOPT_PREFIX).size());
+    }
+    return StringRef(UNOPT_PREFIX + name.str());
+  }
+
+  uint64_t nextPatchpointID() {
+
+  }
+};
+
+} // end anonymous namespace
+
+char CheckPointPass::id = 0;
```

Comment:
Does this stuff need to be global?

[Don't worry too much about fixing this now].

Reply:
These aren't global, they're static member variables.

Reply:
You mean, they are not externed? They are still global in this file, is what I meant. Anyways, don't worry about this. I was thinking out loud mostly.

Reply:
Ah, I see your point. (The whole class is defined inside an anonymous namespace, so it should only be visible in this file.)